src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnRpService.java

```diff
@@ -0,0 +1,241 @@
+        activeRegistrations.put(key, info);
+        return Pair.of(options, key);
+    }
+
+    /**
+     * Returns the username of the user on successful authentication or null if the
+     * authentication was not successful
+     */
+    public Optional<String> finishRegistration(
+            PublicKeyCredential<AuthenticatorAttestationResponse, ClientRegistrationExtensionOutputs> pkc,
+            String realm,
+            String key) {
+        try {
+            final WebAuthnCredentialCreationInfo info = activeRegistrations.get(key);
+            if (info == null || info.realm != realm) {
+                return Optional.empty();
+            }
+            final String username = info.username;
+            final WebAuthnUserAccount account = webAuthnUserAccountRepository.findByProviderAndUsername(provider,
+                    username);
+            assert (account != null);
+            final PublicKeyCredentialCreationOptions options = info.options;
+            RegistrationResult result = rp
+                    .finishRegistration(FinishRegistrationOptions.builder().request(options).response(pkc).build());
+            boolean attestationIsTrusted = result.isAttestationTrusted();
+            if (attestationIsTrusted) {
+                final Set<WebAuthnCredential> previousCredentials = account.getCredentials();
+                final WebAuthnCredential newCred = new WebAuthnCredential();
+                newCred.setCreatedOn(new Date());
+                newCred.setLastUsedOn(new Date());
+                newCred.setCredentialId(result.getKeyId().getId());
+                newCred.setPublicKeyCose(result.getPublicKeyCose());
+                newCred.setSignatureCount(result.getSignatureCount());
+                newCred.setTransports(result.getKeyId().getTransports().orElse(new TreeSet<>()));
+                newCred.setParentAccount(account);
+
+                previousCredentials.add(newCred);
+                webAuthnCredentialsRepository.save(newCred);
+                account.setCredentials(previousCredentials);
+                webAuthnUserAccountRepository.save(account);
+                activeRegistrations.remove(key);
+                return Optional.of(username);
+            }
+        } catch (Exception e) {
+            System.out.println(e);
+        }
+        return Optional.empty();
+    }
+
+    public Pair<AssertionRequest, String> startLogin(String username, String realm) {
+        WebAuthnUserAccount account = webAuthnUserAccountRepository.findByProviderAndUsername(provider, username);
+        StartAssertionOptions startAssertionOptions = StartAssertionOptions.builder()
+                .userHandle(account.getUserHandle()).timeout(TIMEOUT)
+                .userVerification(UserVerificationRequirement.REQUIRED).username(username).build();
+        AssertionRequest startAssertion = rp.startAssertion(startAssertionOptions);
+        final String key = generateNewKey();
```

Comment:
why keys are generated in this method instead of being passed as parameters?

Reply:
Solved with 9a53d7b7d145fe5eb8309cca2f9fe171746b8a90